### PR TITLE
"Connect to Postgres" using ConfigureEmulator API 

### PR DESCRIPTION
**Pull request by hlshen**
* Add UI for Connect to Postgres, and displaying postgres connection after
* Local endpoint is a class function in EmulatorController
* Implement DataConnectEmulatorClient, and ConfigureEmulator api

### CODE REVIEW

File: firebase-vscode/src/core/emulators.ts [hunk 1]
```diff
@@ -241,6 +241,21 @@ export class EmulatorsController implements Disposable {
...
+    const dataConnectEmulator = emulatorInfos?.find(
+      (emulatorInfo) => emulatorInfo.name === Emulators.DATACONNECT,
+    );
+
+    if (!dataConnectEmulator) {
+      return undefined;
+    }
+
+    return (
+      "http://" + dataConnectEmulator.host + ":" + dataConnectEmulator.port
```
Comment (yuchenshi) [inline firebase-vscode/src/core/emulators.ts:255]: Please handle IPv6 quoting similar to https://github.com/firebase/firebase-tools/blob/1633d690e9ed84f9bfe68455fe76a3135681f0b7/src/emulator/registry.ts#L180-L181

File: src/emulator/dataconnectEmulator.ts [hunk 2]
```diff
@@ -83,6 +90,11 @@ export class DataConnectEmulator implements EmulatorInstance {
       timeout: 10_000,
     };
   }
+
+  getOrigin(): string {
```
Comment (yuchenshi) [inline src/emulator/dataconnectEmulator.ts:94]: Where is this used? Do we need it?

File: src/emulator/dataconnectEmulator.ts [hunk 3]
```diff
@@ -30,7 +31,13 @@ export interface DataConnectBuildArgs {
 }
 
 export class DataConnectEmulator implements EmulatorInstance {
-  constructor(private args: DataConnectEmulatorArgs) {}
+  private emulatorClient: DataConnectEmulatorClient;
+  private origin: string;
+
+  constructor(private args: DataConnectEmulatorArgs) {
+    this.origin = args.listen[0].address + args.listen[0].port;
```
Comment (yuchenshi) [inline src/emulator/dataconnectEmulator.ts:38]: Don't manually compute from address and ports, use `EmulatorRegistry.client` instead. `args.listen[0]` doesn't handle `0.0.0.0` and so on correctly, nor does it quote IPv6 properly.

https://github.com/firebase/firebase-tools/blob/1633d690e9ed84f9bfe68455fe76a3135681f0b7/src/emulator/registry.ts#L194

**Review by yuchenshi — APPROVED** [review]
Mostly looking good but please address comments inline

File: src/emulator/dataconnectEmulator.ts [hunk 4]
```diff
@@ -138,4 +150,41 @@ export class DataConnectEmulator implements EmulatorInstance {
...
+  // overwritten.
+  database?: string;
+};
+
+export class DataConnectEmulatorClient {
+  private readonly client: Client;
+  constructor(origin: string) {
+    this.client = new Client({
+      urlPrefix: origin,
+      apiVersion: DownloadDetails.dataconnect.version,
```
[the remark this answers is not in the document]
Comment (hlshen) [reply src/emulator/dataconnectEmulator.ts:176]: ohhhh

File: src/emulator/dataconnectEmulator.ts [hunk 5]
```diff
@@ -138,4 +150,41 @@ export class DataConnectEmulator implements EmulatorInstance {
...
+      auth: false,
+    });
+  }
+
+  public async configureEmulator(body: ConfigureEmulatorRequest) {
+    const res = await this.client.post<ConfigureEmulatorRequest, {} | FirebaseError>(
```
[the remark this answers is not in the document]
Comment (hlshen) [reply src/emulator/dataconnectEmulator.ts:182]: successful return is void. i wasn't sure the class for a typical http error.

File: src/emulator/dataconnectEmulator.ts (hunk 2 above)
Comment (hlshen) [reply src/emulator/dataconnectEmulator.ts:94]: removed

File: src/emulator/dataconnectEmulator.ts (hunk 3 above)
Comment (hlshen) [reply src/emulator/dataconnectEmulator.ts:38]: oh this is a lot smarter, thanks

**Comment by joehan** [comment]
Merged my PR onto yours to autoconnect when not in vscode

File: src/emulator/dataconnectEmulator.ts [hunk 6]
```diff
@@ -2,13 +2,17 @@ import * as childProcess from "child_process";
 
 import { dataConnectLocalConnString } from "../api";
 import { Constants } from "./constants";
-import { getPID, start, stop, downloadIfNecessary } from "./downloadableEmulators";
+import { getPID, start, stop, downloadIfNecessary
```
Comment (joehan) [inline src/emulator/dataconnectEmulator.ts:5]: Looks like we just need to run 'npm run format'

**Review by joehan — APPROVED** [review]
LGTM after formatting

File: firebase-vscode/src/data-connect/emulator.ts [hunk 7]
```diff
@@ -34,40 +33,44 @@ export class DataConnectEmulatorController implements vscode.Disposable {
...
     });
 
     return connectionString;
   }
 
   private async connectToPostgres() {
     const rc = firebaseRC.value?.tryReadValue;
+    const newConnectionString = await this.promptConnectionString(
+      rc?.getDataconnect()?.postgres.localConnectionString ||
+        "postgres://user:password@localhost:5432/dbname",
```
[the remark this answers is not in the document]
Comment (yuchenshi) [reply firebase-vscode/src/data-connect/emulator.ts:53]: yes

**Pull request merged**